src/envoy/http/path_rewrite/filter.cc

```diff
@@ -0,0 +1,120 @@
+namespace path_rewrite {
+
+using Envoy::Http::FilterDataStatus;
+using Envoy::Http::FilterHeadersStatus;
+using Envoy::Http::FilterTrailersStatus;
+using Envoy::Http::RequestHeaderMap;
+
+FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool) {
+  if (headers.Path() == nullptr) {
+    // NOTE: this shouldn't happen in practice because ServiceControl filter
+    // would have already rejected the request.
+    config_->stats().denied_by_no_path_.inc();
+    rejectRequest(Envoy::Http::Code::BadRequest, "No path in request headers",
+                  utils::generateRcDetails(utils::kRcDetailFilterPathRewrite,
+                                           utils::kRcDetailErrorTypeBadRequest,
+                                           utils::kRcDetailErrorMissingPath));
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  absl::string_view original_path = headers.Path()->value().getStringView();
+  // Reject requests with fragment identifiers. They should never be sent to
+  // servers, and it breaks how we handle path translation (query params
+  // appended incorrectly).
+  if (absl::StrContains(original_path, "#")) {
+    config_->stats().denied_by_invalid_path_.inc();
+    rejectRequest(
+        Envoy::Http::Code::BadRequest,
+        "Path cannot contain fragment identifier (#)",
+        utils::generateRcDetails(utils::kRcDetailFilterPathRewrite,
+                                 utils::kRcDetailErrorTypeBadRequest,
+                                 utils::kRcDetailErrorFragmentIdentifier));
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  // Make sure route is calculated
+  auto route = decoder_callbacks_->route();
+  if (route == nullptr || route->routeEntry() == nullptr) {
+    config_->stats().denied_by_no_route_.inc();
+
+    rejectRequest(
+        Envoy::Http::Code::NotFound,
+        absl::StrCat("Request `", utils::readHeaderEntry(headers.Method()), " ",
+                     utils::readHeaderEntry(headers.Path()),
+                     "` is not defined by this API."),
+        utils::generateRcDetails(utils::kRcDetailFilterPathRewrite,
+                                 utils::kRcDetailErrorTypeUndefinedRequest));
+    return FilterHeadersStatus::StopIteration;
+  }
+
+  const auto* per_route =
+      route->routeEntry()->perFilterConfigTyped<PerRouteFilterConfig>(
+          kFilterName);
+  if (per_route == nullptr) {
+    ENVOY_LOG(debug, "no per-route path_rewrite config");
+    config_->stats().path_not_changed_.inc();
+    return FilterHeadersStatus::Continue;
+  }
+
+  std::string new_path;
+  if (!per_route->config_parser().rewrite(original_path, new_path)) {
+    config_->stats().denied_by_url_template_mismatch_.inc();
+    rejectRequest(
+        Envoy::Http::Code::InternalServerError,
+        absl::StrCat("Request `", utils::readHeaderEntry(headers.Method()), " ",
+                     utils::readHeaderEntry(headers.Path()),
+                     "` is mismatched with url_template: ",
```

Comment:
This shouldn't happen. We use the url template to generate the route regex but now the request matched route regex doesn't match url template. It means we have bug in generating route regex(most likely). Could you add the comment here?

We shouldn't specify the url template. For downstream especially for api gateway users, they shouldn't know url template so it will confuse them. Please use other error message and also use new rc details detailed error instead of 
kRcDetailErrorTypeUndefinedRequest for differentiating for upstream.

Reply:
This is really for sanity checking if our Regex in RouteMatch works as intended. I even think about building a path_matcher_sanity envoy filter, to add method + url_template to verify it.  If it doesn't match,  reject the request to expose the problem.

Reply:
I see, the detail error should be in the rc_detail. will change

Reply:
Makes sense, LGTM